File: interoperability/bns019/python/bns019_validator.py
#!/usr/bin/env python3
"""Independent, standard-library-only BNS-019 producer validator.

This implementation intentionally does not import ``bionexus``.  It verifies
the language-neutral release manifest, reads the public registry, and executes
the published producer conformance cases.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import re
import sys
from pathlib import Path, PurePosixPath
from typing import Any, Mapping

MANIFEST_SCHEMA = "urn:bionexus:scientific-semantic-release-manifest:1"
RESULT_SCHEMA = "urn:bionexus:bns019-implementation-result:1"
STANDARD_ID = "BNS-019"
ARTIFACT_NAME = "bionexus-scientific-semantic-conventions"
ATTRIBUTE_NAME = re.compile(r"^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)+$")


class ValidationError(RuntimeError):
    """Release or suite failure that prevents a conformance conclusion."""


def canonical_json(value: Any) -> bytes:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def safe_relative_path(raw: str) -> PurePosixPath:
    path = PurePosixPath(raw)
    if path.is_absolute() or not path.parts or "." in path.parts or ".." in path.parts:
        raise ValidationError(f"unsafe manifest path: {raw!r}")
    return path


def read_object(path: Path, label: str) -> dict[str, Any]:
    try:
        value = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise ValidationError(f"cannot read {label}: {exc}") from exc
    if not isinstance(value, dict):
        raise ValidationError(f"{label} must be a JSON object")
    return value


def load_verified_release(root: Path) -> tuple[dict[str, Any], dict[str, Any]]:
    root = root.resolve()
    manifest = read_object(root / "release-manifest.json", "release manifest")
    if manifest.get("schema") != MANIFEST_SCHEMA:
        raise ValidationError("unsupported release manifest schema")
    if manifest.get("standard_id") != STANDARD_ID or manifest.get("artifact_name") != ARTIFACT_NAME:
        raise ValidationError("release identity mismatch")
    version_path = root / "VERSION"
    try:
        version = version_path.read_text(encoding="utf-8").strip()
    except OSError as exc:
        raise ValidationError(f"cannot read VERSION: {exc}") from exc
    if not version or manifest.get("version") != version:
        raise ValidationError("release version mismatch")

    records = manifest.get("files")
    if not isinstance(records, list) or not records:
        raise ValidationError("manifest files must be a non-empty array")
    seen: set[str] = set()
    for record in records:
        if not isinstance(record, dict):
            raise ValidationError("manifest file record must be an object")
        relative = safe_relative_path(str(record.get("path", "")))
        relative_text = relative.as_posix()
        if relative_text in seen:
            raise ValidationError(f"duplicate manifest path: {relative_text}")
        seen.add(relative_text)
        path = root.joinpath(*relative.parts)
        if not path.is_file():
            raise ValidationError(f"manifest file is missing: {relative_text}")
        if record.get("sha256") != sha256_file(path):
            raise ValidationError(f"SHA-256 mismatch: {relative_text}")
        if record.get("size_bytes") != path.stat().st_size:
            raise ValidationError(f"size mismatch: {relative_text}")

    distributed = {
        path.relative_to(root).as_posix()
        for path in root.rglob("*")
        if path.is_file() and path.name != "release-manifest.json" and "__pycache__" not in path.parts
    }
    if seen != distributed:
        raise ValidationError(
            f"manifest inventory mismatch; unlisted={sorted(distributed - seen)}, absent={sorted(seen - distributed)}"
        )
    expected_digest = hashlib.sha256(canonical_json(records)).hexdigest()
    if manifest.get("release_digest_sha256") != expected_digest:
        raise ValidationError("release_digest_sha256 mismatch")

    registry = read_object(root / "registry.json", "registry")
    if registry.get("schema_version") != version:
        raise ValidationError("registry version mismatch")
    return manifest, registry


def _failure_class(message: str) -> str:
    if " is blocked:" in message:
        return "blocked_legacy_value"
    if message.startswith("missing required attribute:"):
        return "missing_required_attribute"
    if message.startswith("unknown attribute "):
        return "unknown_attribute"
    if message.startswith("unknown value for "):
        return "unknown_registered_value"
    if message.startswith("conflicting values supplied for "):
        return "conflicting_alias"
    if message.startswith("unknown convention group:"):
        return "unknown_convention"
    if " must be " in message or " must contain " in message:
        return "type_or_cardinality"
    return "semantic_validation_error"


def validate_attributes(registry: Mapping[str, Any], convention: str, attributes: Mapping[str, Any]) -> dict[str, Any]:
    groups = registry.get("groups", {})
    definitions = registry.get("attributes", {})
    if convention not in groups:
        errors = [f"unknown convention group: {convention}"]
        return _report({}, errors, [])
    if not isinstance(attributes, Mapping):
        return _report({}, ["attributes must be an object"], [])

    aliases = registry.get("attribute_aliases", {})
    canonical: dict[str, Any] = {}
    errors: list[str] = []
    warnings: list[str] = []
    for supplied_name, supplied_value in attributes.items():
        name = str(supplied_name)
        target = str(aliases.get(name, name))
        if target in canonical and canonical[target] != supplied_value:
            errors.append(f"conflicting values supplied for {target} through an alias")
            continue
        canonical[target] = supplied_value

    extension_pattern = re.compile(str(registry.get("extension_namespace_pattern", "")))
    normalized: dict[str, Any] = {}
    for name in sorted(canonical):
        value = canonical[name]
        if name not in definitions:
            if extension_pattern.fullmatch(name):
                if isinstance(value, str):
                    normalized[name] = value
                elif isinstance(value, list) and all(isinstance(item, str) for item in value):
                    normalized[name] = sorted(set(value))
                else:
                    errors.append(f"extension attribute {name} must be a string or string array")
            else:
                errors.append(f"unknown attribute {name}; custom attributes must use x.<vendor>.*")
            continue

        definition = definitions[name]
        many = definition.get("cardinality") == "many"
        if many:
            if not isinstance(value, list):
                errors.append(f"{name} must be a string array")
                continue
            if not value:
                errors.append(f"{name} must contain at least one value")
                continue
            if not all(isinstance(item, str) for item in value):
                errors.append(f"{name} must contain only strings")
                continue
            values = value
        else:
            if not isinstance(value, str):
                errors.append(f"{name} must be a string")
                continue
            values = [value]

        allowed = set(definition.get("values", []))
        value_aliases = registry.get("value_aliases", {}).get(name, {})
        blocked = registry.get("blocked_legacy_values", {}).get(name, {})
        output: list[str] = []
        item_errors: list[str] = []
        for raw in values:
            if raw in blocked:
                item_errors.append(f"{name}={raw!r} is blocked: {blocked[raw]}")
                continue
            item = str(value_aliases.get(raw, raw))
            if item not in allowed:
                item_errors.append(f"unknown value for {name}: {item!r}")
                continue
            output.append(item)
        errors.extend(item_errors)
        if not item_errors:
            normalized[name] = sorted(set(output)) if many else output[0]

    requirements = groups[convention].get("attributes", {})
    for name, requirement in requirements.items():
        if requirement == "required" and name not in normalized:
            errors.append(f"missing required attribute: {name}")
        elif requirement == "recommended" and name not in normalized:
            warnings.append(f"missing recommended attribute: {name}")
    return _report(normalized, errors, warnings)


def _report(normalized: dict[str, Any], errors: list[str], warnings: list[str]) -> dict[str, Any]:
    failure_classes = list(dict.fromkeys(_failure_class(message) for message in errors))
    return {
        "valid": not errors,
        "normalized_attributes": normalized,
        "failure_classes": failure_classes,
        "errors": errors,
        "warnings": warnings,
    }


def run_conformance_suite(standard_root: Path) -> dict[str, Any]:
    manifest, registry = load_verified_release(standard_root)
    conformance_root = standard_root.resolve() / "conformance"
    suite = read_object(conformance_root / "manifest.json", "conformance manifest")
    case_results: list[dict[str, Any]] = []
    for case in suite.get("cases", []):
        fixture = read_object(conformance_root / str(case["input"]), f"case {case['id']}")
        observed = validate_attributes(registry, fixture.get("convention", ""), fixture.get("attributes", {}))
        matched = observed["valid"] is case["expected_valid"]
        if case["expected_valid"]:
            matched = matched and observed["normalized_attributes"] == case["expected_normalized_attributes"]
        else:
            matched = matched and case["expected_failure_class"] in observed["failure_classes"]
        case_results.append(
            {
                "case_id": case["id"],
                "status": "PASS" if matched else "FAIL",
                "expected_valid": case["expected_valid"],
                "observed_valid": observed["valid"],
                "normalized_attributes": observed["normalized_attributes"],
                "failure_classes": observed["failure_classes"],
            }
        )
    status = "PASS" if case_results and all(case["status"] == "PASS" for case in case_results) else "FAIL"
    return {
        "schema": RESULT_SCHEMA,
        "implementation": {
            "id": "bns019-python-stdlib",
            "track": "independent_validator",
            "language": "python",
        },
        "standard": {
            "id": STANDARD_ID,
            "version": manifest["version"],
            "release_digest_sha256": manifest["release_digest_sha256"],
        },
        "status": status,
        "case_results": case_results,
        "claim_boundary": "Software-contract conformance only; not certification or biological validation.",
    }


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--standard-root", type=Path, required=True)
    parser.add_argument("--output", type=Path)
    args = parser.parse_args(argv)
    try:
        result = run_conformance_suite(args.standard_root)
    except ValidationError as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        return 2
    payload = json.dumps(result, indent=2, ensure_ascii=False) + "\n"
    if args.output:
        args.output.write_text(payload, encoding="utf-8", newline="\n")
    else:
        print(payload, end="")
    return 0 if result["status"] == "PASS" else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: interoperability/bns019/r/bns019_validator.R
#!/usr/bin/env Rscript

# Independent BNS-019 producer validator. This file does not call Python or
# import the BioNexus package; it consumes the same language-neutral release.

if (!requireNamespace("jsonlite", quietly = TRUE)) {
  stop("jsonlite is required", call. = FALSE)
}
if (!requireNamespace("digest", quietly = TRUE)) {
  stop("digest is required for SHA-256 release verification", call. = FALSE)
}

MANIFEST_SCHEMA <- "urn:bionexus:scientific-semantic-release-manifest:1"
RESULT_SCHEMA <- "urn:bionexus:bns019-implementation-result:1"
STANDARD_ID <- "BNS-019"
ARTIFACT_NAME <- "bionexus-scientific-semantic-conventions"

read_object <- function(path, label) {
  value <- tryCatch(
    jsonlite::fromJSON(path, simplifyVector = FALSE),
    error = function(e) stop(sprintf("cannot read %s: %s", label, e$message), call. = FALSE)
  )
  if (!is.list(value) || is.null(names(value))) {
    stop(sprintf("%s must be a JSON object", label), call. = FALSE)
  }
  value
}

sha256_file <- function(path) {
  digest::digest(file = path, algo = "sha256", serialize = FALSE)
}

sha256_text <- function(value) {
  digest::digest(enc2utf8(value), algo = "sha256", serialize = FALSE)
}

canonical_records_json <- function(records) {
  ordered <- lapply(records, function(record) {
    list(path = record$path, sha256 = record$sha256, size_bytes = record$size_bytes)
  })
  as.character(jsonlite::toJSON(
    ordered,
    auto_unbox = TRUE,
    null = "null",
    na = "null",
    digits = NA,
    pretty = FALSE
  ))
}

safe_relative_path <- function(path) {
  if (!is.character(path) || length(path) != 1L || !nzchar(path) ||
      grepl("^(/|[A-Za-z]:)", path) ||
      any(strsplit(gsub("\\\\", "/", path), "/", fixed = TRUE)[[1]] %in% c(".", ".."))) {
    stop(sprintf("unsafe manifest path: %s", path), call. = FALSE)
  }
  path
}

load_verified_release <- function(root) {
  root <- normalizePath(root, winslash = "/", mustWork = TRUE)
  manifest <- read_object(file.path(root, "release-manifest.json"), "release manifest")
  if (!identical(manifest$schema, MANIFEST_SCHEMA)) stop("unsupported release manifest schema", call. = FALSE)
  if (!identical(manifest$standard_id, STANDARD_ID) || !identical(manifest$artifact_name, ARTIFACT_NAME)) {
    stop("release identity mismatch", call. = FALSE)
  }
  version <- trimws(paste(readLines(file.path(root, "VERSION"), warn = FALSE, encoding = "UTF-8"), collapse = "\n"))
  if (!nzchar(version) || !identical(manifest$version, version)) stop("release version mismatch", call. = FALSE)
  records <- manifest$files
  if (!is.list(records) || length(records) == 0L) stop("manifest files must be a non-empty array", call. = FALSE)

  seen <- character()
  for (record in records) {
    relative <- safe_relative_path(record$path)
    if (relative %in% seen) stop(sprintf("duplicate manifest path: %s", relative), call. = FALSE)
    seen <- c(seen, relative)
    path <- do.call(file.path, c(list(root), as.list(strsplit(relative, "/", fixed = TRUE)[[1]])))
    if (!file.exists(path) || dir.exists(path)) stop(sprintf("manifest file is missing: %s", relative), call. = FALSE)
    if (!identical(record$sha256, sha256_file(path))) stop(sprintf("SHA-256 mismatch: %s", relative), call. = FALSE)
    size <- unname(file.info(path)$size)
    if (!identical(as.numeric(record$size_bytes), as.numeric(size))) stop(sprintf("size mismatch: %s", relative), call. = FALSE)
  }
  distributed <- list.files(root, recursive = TRUE, full.names = FALSE, all.files = TRUE, include.dirs = FALSE)
  distributed <- gsub("\\\\", "/", distributed)
  distributed <- distributed[basename(distributed) != "release-manifest.json" & !grepl("(^|/)__pycache__(/|$)", distributed)]
  if (!identical(sort(unique(seen)), sort(unique(distributed)))) stop("manifest inventory mismatch", call. = FALSE)
  release_digest <- sha256_text(canonical_records_json(records))
  if (!identical(manifest$release_digest_sha256, release_digest)) stop("release_digest_sha256 mismatch", call. = FALSE)

  registry <- read_object(file.path(root, "registry.json"), "registry")
  if (!identical(registry$schema_version, version)) stop("registry version mismatch", call. = FALSE)
  list(root = root, manifest = manifest, registry = registry)
}

failure_class <- function(message) {
  if (grepl(" is blocked:", message, fixed = TRUE)) return("blocked_legacy_value")
  if (startsWith(message, "missing required attribute:")) return("missing_required_attribute")
  if (startsWith(message, "unknown attribute ")) return("unknown_attribute")
  if (startsWith(message, "unknown value for ")) return("unknown_registered_value")
  if (startsWith(message, "conflicting values supplied for ")) return("conflicting_alias")
  if (startsWith(message, "unknown convention group:")) return("unknown_convention")
  if (grepl(" must be ", message, fixed = TRUE) || grepl(" must contain ", message, fixed = TRUE)) {
    return("type_or_cardinality")
  }
  "semantic_validation_error"
}

make_report <- function(normalized, errors, warnings) {
  list(
    valid = length(errors) == 0L,
    normalized_attributes = normalized,
    failure_classes = unique(vapply(errors, failure_class, character(1))),
    errors = errors,
    warnings = warnings
  )
}

validate_attributes <- function(registry, convention, attributes) {
  groups <- registry$groups
  definitions <- registry$attributes
  if (is.null(groups[[convention]])) {
    return(make_report(list(), sprintf("unknown convention group: %s", convention), character()))
  }
  if (!is.list(attributes) || is.null(names(attributes))) {
    return(make_report(list(), "attributes must be an object", character()))
  }

  aliases <- registry$attribute_aliases
  canonical <- list()
  errors <- character()
  warnings <- character()
  for (supplied_name in names(attributes)) {
    target <- if (!is.null(aliases[[supplied_name]])) aliases[[supplied_name]] else supplied_name
    supplied_value <- attributes[[supplied_name]]
    if (!is.null(canonical[[target]]) && !identical(canonical[[target]], supplied_value)) {
      errors <- c(errors, sprintf("conflicting values supplied for %s through an alias", target))
    } else {
      canonical[[target]] <- supplied_value
    }
  }

  normalized <- list()
  for (name in sort(names(canonical))) {
    value <- canonical[[name]]
    definition <- definitions[[name]]
    if (is.null(definition)) {
      if (grepl(registry$extension_namespace_pattern, name, perl = TRUE)) {
        if (is.character(value) && length(value) == 1L) {
          normalized[[name]] <- value
        } else if (is.list(value) && all(vapply(value, function(x) is.character(x) && length(x) == 1L, logical(1)))) {
          normalized[[name]] <- as.list(sort(unique(unlist(value, use.names = FALSE))))
        } else {
          errors <- c(errors, sprintf("extension attribute %s must be a string or string array", name))
        }
      } else {
        errors <- c(errors, sprintf("unknown attribute %s; custom attributes must use x.<vendor>.*", name))
      }
      next
    }

    many <- identical(definition$cardinality, "many")
    if (many) {
      if (!is.list(value)) {
        errors <- c(errors, sprintf("%s must be a string array", name))
        next
      }
      if (length(value) == 0L) {
        errors <- c(errors, sprintf("%s must contain at least one value", name))
        next
      }
      if (!all(vapply(value, function(x) is.character(x) && length(x) == 1L, logical(1)))) {
        errors <- c(errors, sprintf("%s must contain only strings", name))
        next
      }
      values <- unlist(value, use.names = FALSE)
    } else {
      if (!is.character(value) || length(value) != 1L) {
        errors <- c(errors, sprintf("%s must be a string", name))
        next
      }
      values <- value
    }

    allowed <- unlist(definition$values, use.names = FALSE)
    value_aliases <- registry$value_aliases[[name]]
    blocked <- registry$blocked_legacy_values[[name]]
    output <- character()
    item_errors <- character()
    for (raw in values) {
      if (!is.null(blocked) && !is.null(blocked[[raw]])) {
        item_errors <- c(item_errors, sprintf("%s='%s' is blocked: %s", name, raw, blocked[[raw]]))
        next
      }
      item <- if (!is.null(value_aliases) && !is.null(value_aliases[[raw]])) value_aliases[[raw]] else raw
      if (!(item %in% allowed)) {
        item_errors <- c(item_errors, sprintf("unknown value for %s: '%s'", name, item))
      } else {
        output <- c(output, item)
      }
    }
    errors <- c(errors, item_errors)
    if (length(item_errors) == 0L) {
      normalized[[name]] <- if (many) as.list(sort(unique(output))) else output[[1]]
    }
  }

  requirements <- groups[[convention]]$attributes
  for (name in names(requirements)) {
    requirement <- requirements[[name]]
    if (identical(requirement, "required") && is.null(normalized[[name]])) {
      errors <- c(errors, sprintf("missing required attribute: %s", name))
    } else if (identical(requirement, "recommended") && is.null(normalized[[name]])) {
      warnings <- c(warnings, sprintf("missing recommended attribute: %s", name))
    }
  }
  make_report(normalized, errors, warnings)
}

canonicalize_json <- function(value) {
  if (!is.list(value)) return(value)
  if (is.null(names(value))) return(lapply(value, canonicalize_json))
  ordered_names <- sort(names(value))
  output <- lapply(value[ordered_names], canonicalize_json)
  names(output) <- ordered_names
  output
}

json_equal <- function(left, right) {
  identical(
    as.character(jsonlite::toJSON(canonicalize_json(left), auto_unbox = TRUE, null = "null", digits = NA)),
    as.character(jsonlite::toJSON(canonicalize_json(right), auto_unbox = TRUE, null = "null", digits = NA))
  )
}

run_conformance_suite <- function(standard_root) {
  release <- load_verified_release(standard_root)
  conformance_root <- file.path(release$root, "conformance")
  suite <- read_object(file.path(conformance_root, "manifest.json"), "conformance manifest")
  case_results <- list()
  for (case in suite$cases) {
    fixture <- read_object(file.path(conformance_root, case$input), sprintf("case %s", case$id))
    observed <- validate_attributes(release$registry, fixture$convention, fixture$attributes)
    matched <- identical(observed$valid, case$expected_valid)
    if (isTRUE(case$expected_valid)) {
      matched <- matched && json_equal(observed$normalized_attributes, case$expected_normalized_attributes)
    } else {
      matched <- matched && case$expected_failure_class %in% observed$failure_classes
    }
    case_results[[length(case_results) + 1L]] <- list(
      case_id = case$id,
      status = if (matched) "PASS" else "FAIL",
      expected_valid = case$expected_valid,
      observed_valid = observed$valid,
      normalized_attributes = observed$normalized_attributes,
      failure_classes = as.list(observed$failure_classes)
    )
  }
  status <- if (length(case_results) > 0L && all(vapply(case_results, function(x) identical(x$status, "PASS"), logical(1)))) "PASS" else "FAIL"
  list(
    schema = RESULT_SCHEMA,
    implementation = list(id = "bns019-r-jsonlite", track = "independent_validator", language = "r"),
    standard = list(
      id = STANDARD_ID,
      version = release$manifest$version,
      release_digest_sha256 = release$manifest$release_digest_sha256
    ),
    status = status,
    case_results = case_results,
    claim_boundary = "Software-contract conformance only; not certification or biological validation."
  )
}

parse_args <- function(args) {
  output <- list(standard_root = NULL, output = NULL)
  index <- 1L
  while (index <= length(args)) {
    if (args[[index]] == "--standard-root" && index < length(args)) {
      output$standard_root <- args[[index + 1L]]
      index <- index + 2L
    } else if (args[[index]] == "--output" && index < length(args)) {
      output$output <- args[[index + 1L]]
      index <- index + 2L
    } else {
      stop(sprintf("unknown or incomplete argument: %s", args[[index]]), call. = FALSE)
    }
  }
  if (is.null(output$standard_root)) stop("--standard-root is required", call. = FALSE)
  output
}

main <- function() {
  args <- parse_args(commandArgs(trailingOnly = TRUE))
  result <- run_conformance_suite(args$standard_root)
  payload <- jsonlite::toJSON(result, auto_unbox = TRUE, null = "null", digits = NA, pretty = TRUE)
  if (is.null(args$output)) {
    cat(payload, "\n", sep = "")
  } else {
    writeLines(enc2utf8(payload), args$output, useBytes = TRUE)
  }
  if (identical(result$status, "PASS")) 0L else 1L
}

if (sys.nframe() == 0L) {
  status <- tryCatch(main(), error = function(e) {
    message("ERROR: ", e$message)
    2L
  })
  quit(save = "no", status = status, runLast = FALSE)
}
